File: python/suada_mcp.py
#!/usr/bin/env python3
"""
Suada Model Context Protocol (MCP) Implementation

This module implements the Model Context Protocol for Suada, allowing models
to retrieve and reason over custom data feeds through Suada's data pipelines.
"""

import os
import json
import logging
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass, field, asdict
import requests
from dotenv import load_dotenv

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger("suada-mcp")

# Load environment variables
load_dotenv()

# MCP Protocol Types
@dataclass
class MCPRequest:
    """Model Context Protocol request structure."""
    query: str
    context: Dict[str, Any] = field(default_factory=dict)
    user_id: Optional[str] = None
    conversation_id: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

@dataclass
class MCPResponse:
    """Model Context Protocol response structure."""
    content: str
    metadata: Dict[str, Any] = field(default_factory=dict)
    error: Optional[str] = None

class SuadaMCP:
    """
    Suada Model Context Protocol implementation.
    
    This class implements the MCP protocol for Suada, allowing models to
    retrieve and reason over custom data feeds through Suada's data pipelines.
    """
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "https://suada.ai/api/public",
        external_user_identifier: Optional[str] = None
    ):
        """
        Initialize the Suada MCP client.
        
        Args:
            api_key: Suada API key. If not provided, will look for SUADA_API_KEY env var.
            base_url: Base URL for Suada API. Defaults to production API.
            external_user_identifier: Default user identifier to use if not specified in requests.
        """
        self.api_key = api_key or os.getenv("SUADA_API_KEY")
        if not self.api_key:
            raise ValueError("Suada API key is required. Provide it as an argument or set SUADA_API_KEY environment variable.")
        
        self.base_url = base_url.rstrip("/")
        self.external_user_identifier = external_user_identifier
        
        logger.info(f"Initialized Suada MCP client with base URL: {self.base_url}")
    
    def _get_headers(self) -> Dict[str, str]:
        """Get HTTP headers for API requests."""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "SuadaMCP/1.0 Python"
        }
    
    def process(self, request: MCPRequest) -> MCPResponse:
        """
        Process an MCP request through Suada.
        
        Args:
            request: The MCP request to process.
            
        Returns:
            An MCPResponse object containing the processed result.
        """
        try:
            # Ensure we have a user identifier
            user_id = request.user_id or self.external_user_identifier
            if not user_id:
                raise ValueError("User identifier is required. Provide it in the request or during initialization.")
            
            # Prepare the payload for Suada API
            payload = {
                "message": request.query,
                "externalUserIdentifier": user_id,
                "context": request.context
            }
            
            # Add conversation ID if available
            if request.conversation_id:
                payload["conversationId"] = request.conversation_id
            
            # Make the API request
            logger.info(f"Sending request to Suada API: {request.query}")
            response = requests.post(
                f"{self.base_url}/chat",
                headers=self._get_headers(),
                json=payload
            )
            
            # Handle API errors
            response.raise_for_status()
            
            # Parse the response
            data = response.json()
            
            # Extract structured data from the response
            structured_data = {
                "metrics": data.get("metrics", {}),
                "insights": data.get("insights", []),
                "recommendations": data.get("recommendations", []),
                "risks": data.get("risks", []),
                "reasoning": data.get("reasoning", "")
            }
            
            # Return formatted response
            return MCPResponse(
                content=data.get("response", ""),
                metadata={
                    "suada_data": structured_data,
                    "raw_response": data
                }
            )
            
        except requests.exceptions.RequestException as e:
            logger.error(f"API request error: {str(e)}")
            
            # Try to extract error details if available
            error_message = "Failed to communicate with Suada API"
            if hasattr(e, "response") and e.response is not None:
                try:
                    error_data = e.response.json()
                    if "message" in error_data:
                        error_message = error_data["message"]
                except:
                    pass
            
            return MCPResponse(
                content="",
                error=error_message,
                metadata={"status_code": e.response.status_code if hasattr(e, "response") else None}
            )
        except Exception as e:
            logger.error(f"Unexpected error: {str(e)}")
            return MCPResponse(
                content="",
                error=f"Unexpected error: {str(e)}",
                metadata={}
            )

    def create_langchain_tool(self, name: str = "suada_data", description: str = None):
        """
        Create a LangChain tool that uses this MCP.
        
        Args:
            name: Name of the tool
            description: Description of the tool
            
        Returns:
            A LangChain tool that can be used in agents
        """
        try:
            # Import LangChain dependencies
            from langchain.tools import BaseTool
            
            default_description = (
                "Use this tool to get business insights and analysis from Suada. "
                "Input should be a specific business question."
            )
            
            # Create a LangChain tool class
            class SuadaTool(BaseTool):
                name = name
                description = description or default_description
                
                def _run(self, query: str) -> str:
                    mcp_request = MCPRequest(query=query)
                    mcp_response = self.mcp.process(mcp_request)
                    
                    if mcp_response.error:
                        return f"Error: {mcp_response.error}"
                    
                    return mcp_response.content
                
                async def _arun(self, query: str) -> str:
                    # For simplicity, we're using the sync version
                    return self._run(query)
            
            # Create an instance with reference to this MCP
            tool_instance = SuadaTool()
            tool_instance.mcp = self
            
            return tool_instance
            
        except ImportError:
            logger.error("LangChain is not installed. Install it with 'pip install langchain'")
            raise ImportError("LangChain is required to create a LangChain tool. Install it with 'pip install langchain'")

# Example usage
if __name__ == "__main__":
    # Initialize the MCP
    mcp = SuadaMCP(
        api_key=os.getenv("SUADA_API_KEY"),
        external_user_identifier="example_user"
    )
    
    # Create a request
    request = MCPRequest(
        query="What's our revenue trend for the last quarter?",
        context={"timeframe": "last_quarter"},
        user_id="example_user"
    )
    
    # Process the request
    response = mcp.process(request)
    
    # Print the response
    if response.error:
        print(f"Error: {response.error}")
    else:
        print(f"Response: {response.content}")
        print(f"Metrics: {response.metadata['suada_data']['metrics']}")
        print(f"Insights: {response.metadata['suada_data']['insights']}") 

File: typescript/suada-mcp.ts
/**
 * Suada Model Context Protocol (MCP) Implementation
 * 
 * This module implements the Model Context Protocol for Suada, allowing models
 * to retrieve and reason over custom data feeds through Suada's data pipelines.
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios';

/**
 * MCP Request interface
 */
export interface MCPRequest {
    query: string;
    context?: Record<string, any>;
    userId?: string;
    conversationId?: string;
    metadata?: Record<string, any>;
}

/**
 * MCP Response interface
 */
export interface MCPResponse {
    content: string;
    metadata: Record<string, any>;
    error?: string;
}

/**
 * Suada API Response structure
 */
export interface SuadaResponse {
    metrics?: Record<string, string>;
    insights?: string[];
    recommendations?: string[];
    risks?: string[];
    reasoning?: string;
    response: string;
}

/**
 * Suada MCP Configuration
 */
export interface SuadaMCPConfig {
    apiKey: string;
    baseUrl?: string;
    externalUserIdentifier?: string;
}

/**
 * Suada Model Context Protocol implementation
 */
export class SuadaMCP {
    private apiKey: string;
    private baseUrl: string;
    private externalUserIdentifier?: string;
    private client: AxiosInstance;

    /**
     * Initialize the Suada MCP client
     * 
     * @param config - Configuration for the Suada MCP client
     */
    constructor(config: SuadaMCPConfig) {
        this.apiKey = config.apiKey;

        if (!this.apiKey) {
            throw new Error('Suada API key is required');
        }

        this.baseUrl = config.baseUrl?.replace(/\/$/, '') || 'https://suada.ai/api/public';
        this.externalUserIdentifier = config.externalUserIdentifier;

        // Initialize axios client
        this.client = axios.create({
            baseURL: this.baseUrl,
            headers: {
                'Authorization': `Bearer ${this.apiKey}`,
                'Content-Type': 'application/json',
                'Accept': 'application/json',
                'User-Agent': 'SuadaMCP/1.0 TypeScript'
            }
        });

        console.log(`Initialized Suada MCP client with base URL: ${this.baseUrl}`);
    }

    /**
     * Process an MCP request through Suada
     * 
     * @param request - The MCP request to process
     * @returns A promise that resolves to an MCPResponse
     */
    async process(request: MCPRequest): Promise<MCPResponse> {
        try {
            // Ensure we have a user identifier
            const userId = request.userId || this.externalUserIdentifier;
            if (!userId) {
                throw new Error('User identifier is required. Provide it in the request or during initialization.');
            }

            // Prepare the payload for Suada API
            const payload = {
                message: request.query,
                externalUserIdentifier: userId,
                context: request.context || {}
            } as Record<string, any>;

            // Add conversation ID if available
            if (request.conversationId) {
                payload.conversationId = request.conversationId;
            }

            // Make the API request
            console.log(`Sending request to Suada API: ${request.query}`);
            const response = await this.client.post<SuadaResponse>('/chat', payload);

            // Extract data from the response
            const data = response.data;

            // Extract structured data
            const structuredData = {
                metrics: data.metrics || {},
                insights: data.insights || [],
                recommendations: data.recommendations || [],
                risks: data.risks || [],
                reasoning: data.reasoning || ''
            };

            // Return formatted response
            return {
                content: data.response || '',
                metadata: {
                    suadaData: structuredData,
                    rawResponse: data
                }
            };

        } catch (error) {
            console.error('Error processing MCP request:', error);

            // Handle axios errors
            if (axios.isAxiosError(error)) {
                const axiosError = error as AxiosError;

                // Try to extract error details if available
                let errorMessage = 'Failed to communicate with Suada API';
                let statusCode = axiosError.response?.status;

                if (axiosError.response?.data && typeof axiosError.response.data === 'object') {
                    const errorData = axiosError.response.data as any;
                    if (errorData.message) {
                        errorMessage = errorData.message;
                    }
                }

                return {
                    content: '',
                    error: errorMessage,
                    metadata: { statusCode }
                };
            }

            // Handle other errors
            return {
                content: '',
                error: `Unexpected error: ${(error as Error).message || String(error)}`,
                metadata: {}
            };
        }
    }

    /**
     * Create a LangChain tool that uses this MCP
     * 
     * @param name - Name of the tool
     * @param description - Description of the tool
     * @returns A LangChain tool that can be used in agents
     */
    createLangChainTool(name = 'suada_data', description?: string) {
        try {
            // Check if LangChain is available
            // This is a runtime check since we don't want to make LangChain a required dependency
            const langchain = require('langchain/tools');

            const defaultDescription =
                'Use this tool to get business insights and analysis from Suada. ' +
                'Input should be a specific business question.';

            // Create a tool class
            class SuadaTool extends langchain.Tool {
                name = name;
                description = description || defaultDescription;
                mcp: SuadaMCP;

                constructor(mcp: SuadaMCP) {
                    super();
                    this.mcp = mcp;
                }

                async _call(query: string): Promise<string> {
                    const mcpRequest: MCPRequest = { query };
                    const mcpResponse = await this.mcp.process(mcpRequest);

                    if (mcpResponse.error) {
                        return `Error: ${mcpResponse.error}`;
                    }

                    return mcpResponse.content;
                }
            }

            // Return an instance of the tool
            return new SuadaTool(this);

        } catch (error) {
            console.error('LangChain is not installed. Install it with npm install langchain');
            throw new Error('LangChain is required to create a LangChain tool. Install it with npm install langchain');
        }
    }
}

// Example usage
/*
async function example() {
  // Initialize the MCP
  const mcp = new SuadaMCP({
    apiKey: process.env.SUADA_API_KEY || '',
    externalUserIdentifier: 'example_user'
  });
  
  // Create a request
  const request: MCPRequest = {
    query: "What's our revenue trend for the last quarter?",
    context: { timeframe: 'last_quarter' },
    userId: 'example_user'
  };
  
  // Process the request
  const response = await mcp.process(request);
  
  // Print the response
  if (response.error) {
    console.error(`Error: ${response.error}`);
  } else {
    console.log(`Response: ${response.content}`);
    console.log(`Metrics:`, response.metadata.suadaData.metrics);
    console.log(`Insights:`, response.metadata.suadaData.insights);
  }
}
*/ 